xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,186 @@
+/** @file featurelist_internal.cc
+ * @brief Definition of Feature::Internal class.
+ */
+/* Copyright (C) 2012 Parth Gupta
+ * Copyright (C) 2016 Ayush Tomar
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+
+#include "xapian-letor/featurelist.h"
+#include "featurelist_internal.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include "debuglog.h"
+
+using namespace std;
+using namespace Xapian;
+
+void
+FeatureList::Internal::set_database(const Xapian::Database & db)
+{
+    LOGCALL_VOID(API, "FeatureList::set_database", db);
+    featurelist_db = db;
+}
+
+void
+FeatureList::Internal::set_query(const Xapian::Query & query)
+{
+    LOGCALL_VOID(API, "FeatureList::set_query", query);
+    if (query.empty()) {
+	throw Xapian::InvalidArgumentError("Can't initialise with an empty query string");
+    }
+    featurelist_query = query;
+}
+
+void
+FeatureList::Internal::set_doc(const Xapian::Document & doc)
+{
+    LOGCALL_VOID(API, "FeatureList::set_doc", doc);
+    featurelist_doc = doc;
+}
+
+void
+FeatureList::Internal::compute_statistics(const Xapian::Query & letor_query, const Xapian::Database & letor_db,
+					  const Xapian::Document & letor_doc)
+{
+    set_query(letor_query);
+    set_doc(letor_doc);
+    set_database(letor_db);
+
+    compute_termfreq();
+    compute_inverse_doc_freq();
+    compute_doc_length();
+    compute_collection_length();
+    compute_collection_termfreq();
+}
+
+void
+FeatureList::Internal::compute_termfreq()
+{
+    std::map<std::string, long int> tf;
+
+    Xapian::TermIterator docterms = featurelist_doc.termlist_begin();
+    for (Xapian::TermIterator qt = featurelist_query.get_terms_begin(); qt != featurelist_query.get_terms_end(); ++qt) {
+	docterms.skip_to(*qt);
+	if (docterms != featurelist_doc.termlist_end() && *qt == *docterms) {
+	    tf[*qt] = docterms.get_wdf();
+	} else {
+	    tf[*qt] = 0;
+	}
+    }
+    termfreq = tf;
+}
+
+void
+FeatureList::Internal::compute_inverse_doc_freq()
+{
+    std::map<std::string, double> idf;
+
+    for (Xapian::TermIterator qt = featurelist_query.get_terms_begin(); qt != featurelist_query.get_terms_end(); ++qt) {
+	if (featurelist_db.term_exists(*qt)) {
```

Comment:
Rather than calling `term_exists()` and then `get_termfreq() when it does, it would be more efficient to just call `get_termfreq()` and if it returns zero, you know the term doesn't exist.